### DISCUSSION TITLE
Grepping Brew List now only errors.

### DISCUSSION THREAD

**Discussion by gabeio** [Homebrew/discussions#519] [Everyday usage] [answered]
I am posting it here as I'm not sure this is a bug or an intended change, but I used to be able to grep everything installed via brew with just: `brew list | grep {package}` but a recent update seems to have changed the list functionality. When reading the source code it looks like it _should_ still work how I think it should, where `brew list` defaults to formulas and casks but when piping/grepping list with no flags it just doesn't allow it.

I tried searching the issues and discussions about this, but it seems the only one regarding something similar to this wasn't posted correctly and was closed and ignored. 

```sh
$ brew list | grep anything
Error: Calling `brew list` to only list formulae is disabled! Use `brew list --formula` instead.
```

Question:

Is this a bug?

**Comment by SMillerDev** [comment, marked as answer]
I mean, it says what you need to do. It seems you misread the code and it doesn't allow mixed lists.

**Reply by gabeio** [reply]
> I mean, it says what you need to do. It seems you misread the code and it doesn't allow mixed lists.

How I read that, I would tend to disagree, as `brew list` alone does print both, piping is the only thing that was _apparently_ disabled. But I guess I got my answer which is, no it is not a bug.

**Reply by btoone** [reply]
I think I understand why using the `--formula` flag is preferred but I would suggest that since it now creates different output (a list of file paths) breaking/preventing our ability pipe the output is step backwards.

For example how should we pattern match now if we want to search to see if multiple packages are installed, not just one? Doing `brew list | grep postgres*` errors as well. The desire being to show me all the postgres formulas I have installed.

I discovered the `-1` flag which puts all the formula in a list, so I thought maybe that was the preferred way to make it grepable. Nope, `brew list -1 | grep postgres*` errors in the same manner.

Then I thought ... okay well I'll just redirect it to a file and then I can `grep` the file with `cat`. Nope, `brew list > out.txt` errors in the same manner as well.

Ideas: I would like to see one of: the `list` command updated so its output can be piped again (at least for the `-1` flag), the `--formula` flag updated to support wildcard args, or some other way to query for installed packages and manipulate the output (grepping or out to a file). Thanks.

@gabeio I had the same question (and searched for a while) so thanks for posting.

Cc @norm

**Reply by Rylan12** [reply]
`brew list --formula | grep` will work. So will `brew list --formula > out.txt`

Any time you see the error simply add `--formula`.

I acknowledge that this is an awkward change and I apologize for that. However, it was necessary to not cause widespread breakage while we merged `brew cask <cmd>` into `brew <cmd>`. I wrote a little bit more about the reasoning over in [this discourse thread](https://discourse.brew.sh/t/brew-list-less-fails/9224).

**Reply by btoone** [reply]
Thank you @Rylan12 for pointing that out and linking to the reasoning.

**Reply by gabeio** [reply]
@Rylan12 agreed, thanks for the link. I understand for scripting that `brew list --formula` makes sense, it's just when one sees `brew list` work and then `brew list | grep {}` fail it's not immediately obvious that error message is related/pertinent.

Reasoning: `Calling 'brew list' to only list formulae is disabled!` tells me I shouldn't call brew list to get formulas only, but I _wanted_ both formulas and casks when greping manually, which is what reading the manual *seems* to indicate.

Possibly improved wording on the error message would improve the questions surrounding it. I'll take a look around to see if I can make a PR to help instead of just telling you my thoughts 😄.

**Reply by carlocab** [reply]
Yes, I can see what you mean about the error message. A PR to clarify this would certainly be helpful.

**Reply by tlk** [reply]
> Any time you see the error simply add `--formula`.

I guess it is too late to reconsider this breaking change?

How about leaving the list command as it was and add a `--all`, `--combined` or `--union` option to retrieve the combined list of formulas and casks?

A small workaround for the current situation is to add an alias to ~/.zshrc or ~/.bashrc:
```bash
echo "alias fml='brew list --formula'" >> ~/.zshrc
```

This alias will make it possible to run `fml` instead of `brew list --formula` in new shell sessions:
```bash
$ fml | grep anything
```

**Reply by Rylan12** [reply]
> I guess it is too late to reconsider this breaking change?

Yes, sorry. I'll point out that this change did go through the appropriate deprecation period where a warning was printed telling users that it would soon change.

> How about leaving the list command as it was and add a `--all`, `--combined` or `--union` option to retrieve the combined list of formulas and casks?

"leaving the list command as it was" might not do what you expect anyway. `brew list | grep` _never_ included casks. That's part of the reason we're making this change is so that it can in the future.

I would guess (don't hold me to this as I'm not sure what's being planned) that starting in Homebrew 2.8.0, `brew list` will always list both formulae and casks even to non-tty output (which will allow `brew list | grep` to use formulae and casks). You're welcome to open a PR adding one of those flags. If it's up to me, I'd rather not add a new flag if it's going to be useless with the next major or minor Homebrew release, but other maintainers may not agree with me.

**Reply by tlk** [reply]
Ok. I don't think it makes sense to open PRs unless the breaking change is reverted such that `brew list` never includes casks and such that `brew list | grep` works like it used to.

I guess I didn't use the command during the deprecation period or I may have been conveniently ignoring the warning :-)

Maybe it would be helpful to others if the error message has a link to an explanation of why this breaking change was deemed necessary.
